### PR TITLE
Fix #14: Allows relative imports with subdirectories.

### LINKED ISSUE #14: Schemas that live in sub-namespaces cause imports to fail.
Consider a schema like this:
<details><summary>simple_schema_1.avsc</summary>

```avro
[
    {
        "type": "record",
        "name": "Alpha",
        "namespace": "base.a",
        "fields": [
            {
                "name": "azi",
                "type": "long"
            },
            {
                "name": "abi",
                "type": "base.Beta"
            }

        ]
    },
    {
        "type": "record",
        "name": "Beta",
        "namespace": "base",
        "fields": [
            {
                "name": "bar",
                "type": "string"
            }
        ]
    }
]
```

</details>

running avro2py on this schema results in it making a file structure like:

```
base/__init__.py
base/a.py
```

And, relevantly for this issue, `a.py` looks like this:
<details><summary>a.py</summary>

```python
"""Schema definitions for `base.a` namespace. Generated by avro2py v.0.0.6."""
from . import Beta
from typing import NamedTuple


class Alpha(NamedTuple):
    azi: int
    abi: "Beta"
    _original_schema = (
        '{"type": "record", "name": "Alpha", "namespace": "base.a", "fields": [{"name":'
        ' "azi", "type": "long"}, {"name": "abi", "type": "base.Beta"}]}'
    )
```

</details>

This is fine. However, if you add a namespace beneath `a`, it breaks. For example, take this schema:

<details><summary>simple_schema_2.avsc</summary>

```avro
[
    {
        "type": "record",
        "name": "Alpha",
        "namespace": "base2.a",
        "fields": [
            {
                "name": "azi",
                "type": "long"
            },
            {
                "name": "abi",
                "type": "base2.Beta"
            }

        ]
    },
    {
        "type": "record",
        "name": "Beta",
        "namespace": "base2",
        "fields": [
            {
                "name": "bar",
                "type": "string"
            }
        ]
    },
    {
        "type": "record",
        "name": "Sub",
        "namespace": "base2.a.s",
        "fields": [
            {
                "name": "foo",
                "type": "string"
            }
        ]
    }
]

```

</details>

If you run avro2py on that, you'll get a file structure like:
```
base2/__init__.py
base2/a/__init__.py
base2/a/s.py
```

But notably for this, `base2/a/__init__.py` looks like this:

<details><summary>a/__init__.py</summary>

```python
"""Schema definitions for `base2.a` namespace. Generated by avro2py v.0.0.6."""
from . import Beta
from typing import NamedTuple


class Alpha(NamedTuple):
    azi: int
    abi: "Beta"
    _original_schema = (
        '{"type": "record", "name": "Alpha", "namespace": "base2.a", "fields":'
        ' [{"name": "azi", "type": "long"}, {"name": "abi", "type": "base2.Beta"}]}'
    )

```

</details>

This import is bad so it doesn't work. It tries to import `Beta` from `.` but now that we've moved this file into a subdirectory it no longer has `Beta` in `.`.

### PR DESCRIPTION
**Pull request by jcoates**
If a and b are siblings, and you are importing `foo` from b into a,
before this change, you'd import like `from . import foo`.

However, if a has a child namespace `bar`, then things in a will live in
an `a/__init__.py` file, so the import needs to be `from .. import foo`.

~I fixed this by making the `bridge_namespaces` method no longer
static (as the object it gets called on has the context of other
namespaces) and then checking if the import-to namespace is a parent of
another namespace.~

~If you would rather we pass the namespaces as an argument or a flag we
could do that, but since the only actual call has the object, it felt
reasonable to do it this way (it did make the test a bit more awkward).~

Based off feedback I went with passing the precomputed list of parent namespaces
as an argument to the still static method. Now the `bridge_namespaces` method needs
to be told what namespaces are parents, but the rewriter calculates those on creation.

I have also added a couple of tests.

Fixes #14

Also, this avoids but doesn't really address #16 (by pinning the dependency).

### CODE REVIEW

File: tests/test_rendering.py [hunk 1]
```diff
@@ -46,7 +46,9 @@ def test_example_message_round_trippable(data):
     example = imp.module_from_spec(spec)
     sys.modules['example'] = example
     spec.loader.exec_module(example)
-    original_example_avro_model = data.draw(st.from_type(example.ExampleAvroModel))
+    st.register_type_strategy(Decimal, st.decimals(allow_nan=False, allow_infinity=False, places=2))
```
Comment (hchasestevens) [inline tests/test_rendering.py]: I see a couple issues here:
1. Per https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.register_type_strategy , this appears to register the type strategy globally, rather than just for this test (or one run thereof). This presumably means that other tests attempting to draw `Decimal` are affected, which seems like behavior we want to avoid; alternatively, if we _did_ want this to be overridden globally for the entire test suite, it seems like we'd want it in a `conftest.py` or similar as opposed to buried in a single test.
2. If we limit the space for `Decimal`, I think it's only fair that we remove the now-redundant `assume` calls from https://github.com/hchasestevens/avro2py/pull/15/files#diff-530cd1c39b1adde7f11281b896978de50dae64009a7589c4f3c5ede8c76f1005R55-R60 

Rather than this, can we not modify https://github.com/hchasestevens/avro2py/pull/15/files#diff-530cd1c39b1adde7f11281b896978de50dae64009a7589c4f3c5ede8c76f1005R62-R63 to just set the `decimal` and `maybeDecimal` attributes to draws of `st.decimals(allow_nan=False, allow_infinity=False, places=2)` explicitly?

File: avro2py/codegen.py [hunk 2]
```diff
@@ -363,10 +362,16 @@ def bridge_namespaces(from_namespace: List[str], to_namespace: List[str], target
         remaining_to_components = [c for c in paired_remaining_to_components if c is not None]
 
         if remaining_from_components:
+            level = len(remaining_from_components)
+            for namespace_name in self.namespaces.keys():
```
Comment (hchasestevens) [inline avro2py/codegen.py]:
```suggestion
            for namespace_name in self.namespaces:
```

Comment (hchasestevens) [inline avro2py/codegen.py]: I see that the access to `self.namespaces` means that `bridge_namespaces` is no longer a `@staticmethod` - since `bridge_namespaces` has some pretty complex logic that was _previously_ pure, I would ask that either:
1. We simply pass in these namespaces and revert to being a `@staticmethod`, to make it easier to reason about what this method is doing in future (e.g. we'll know it's not mutating anything), or
2. We extract the pure components of the method (i.e. https://github.com/hchasestevens/avro2py/pull/15/files#diff-dd17a87ea1ee28b0a8ef80f33739b424462dfd309d0a7eb4d58b619f0ec386f3R347-R362 ) into a new `@staticmethod` and write tests to exercise that new method specifically.

File: example/schema.avsc [hunk 3]
```diff
@@ -708,7 +708,7 @@
                 }
               },
               {
-                "name": "value",
+                "name": "ivalue",
```
Comment (hchasestevens) [inline example/schema.avsc]: A couple of things here:
1. You've changed the example schema here, but haven't regenerated the example codegen - we need to keep these in sync.
2. It seems like the change to using `ivalue` and `svalue`, to sidestep the field names colliding, deliberately hides the problem you mention in https://github.com/hchasestevens/avro2py/issues/16 - that seems like a bad solution. I would instead much prefer that you alter `requirements.txt` to specify `avro<={MAX COMPATIBLE VERSION}` so that we are explicit about which versions we're _actually_ compatible with.

File: avro2py/codegen.py [hunk 4]
```diff
@@ -363,10 +362,16 @@ def bridge_namespaces(from_namespace: List[str], to_namespace: List[str], target
...
+                namespace_parts = namespace_name.split('.')
+                if namespace_parts[0:len(from_namespace)] == from_namespace and len(namespace_parts) > len(from_namespace):
+                    level += 1
+                    break
```
Comment (hchasestevens) [inline avro2py/codegen.py]: In your PR description, you say:
> then checking if the import-to namespace is a parent of another namespace.
, but I think the way this is written doesn't make it very clear that that's what's occurring here (meaning - I can't very easily follow the logic and see how this code is implementing what you say).

I would probably say, to begin, that we should either have this be a separate method that's named appropriately (like `check_if_to_namespace_is_parent`) or comments added to this snippet describing what we're doing and why.

I would also say, rather than having this be written as an "exceptional" case where we need to check through all the namespaces to see if any qualify as parents, I would much prefer if we pre-computed this upfront for all namespaces, and store this explicitly as some `namespace_is_parent: Dict[Tuple[str], bool]` or `parent_namespaces: FrozenSet[Tuple[str]]` value upon initialization of the class, which we could then just be interrogating here.

File: avro2py/codegen.py [hunk 5]
```diff
@@ -363,10 +362,16 @@ def bridge_namespaces(from_namespace: List[str], to_namespace: List[str], target
         remaining_to_components = [c for c in paired_remaining_to_components if c is not None]
 
         if remaining_from_components:
+            level = len(remaining_from_components)
+            for namespace_name in self.namespaces.keys():
+                namespace_parts = namespace_name.split('.')
+                if namespace_parts[0:len(from_namespace)] == from_namespace and len(namespace_parts) > len(from_namespace):
```
Comment (hchasestevens) [inline avro2py/codegen.py]: This condition is hard to follow. Would you mind breaking it up like e.g.:
```suggestion
                namespace_is_in_from_namespace = all(
                    namespace_component == from_namespace_component
                    for namespace_component, from_namespace_component in zip(namespace_parts, from_namespace)
                )
                namespace_is_deeper_than_from_namespace = len(namespace_parts) > len(from_namespace_parts)
                namespace_is_child_of_from_namespace = namespace_is_in_from_namespace and namespace_is_deeper_than_from_namespace 
                if namespace_is_child_of_from_namespace:
```

File: avro2py/codegen.py (hunk 4 above)
Comment (hchasestevens) [inline avro2py/codegen.py]: I think that rather than doing:
```python
for x in xs:
    if cond(x):
        do_something_not_requiring_x()
        break
```
this would be more clearly expressed if we did:
```python
if any(cond(x) for x in xs):
    do_something_not_requiring_x()
```
, meaning, in this case,
```python
if any(is_child(parent=from_namespace, target=namespace) for namespace in self.namespaces):
    level = len(remaining_from_components) + 1
else:
    level = len(remaining_from_components)
```
(or, if you implement my other comment, 
```python
if from_namespace in self.namespaces_with_children:
    level = len(remaining_from_components) + 1
else:
    level = len(remaining_from_components)
```

File: avro2py/codegen.py [hunk 6]
```diff
@@ -363,10 +362,16 @@ def bridge_namespaces(from_namespace: List[str], to_namespace: List[str], target
...
+                    level += 1
```
Comment (hchasestevens) [inline avro2py/codegen.py]: The `+= 1` here is strange to me, since we immediately break, and I think supports that this would be better expressed as some method like that's doing an `if cond: return True` rather than `if cond: x(); break`.

**Review by hchasestevens — CHANGES_REQUESTED** [review]

File: avro2py/codegen.py (hunk 5 above)
Comment (hchasestevens) [reply avro2py/codegen.py]: Alternatively, if you just did a check like `if from_namespace == namespace_parts: continue` before the `zip` check, I think the `zip` would be sufficient - you're just trying to see if the `from_namespace` has _any_ children, and need to make sure that you're excluding the trivial case of comparing `from_namespace` to itself, right?

File: tests/test_rendering.py (hunk 1 above)
Comment (jcoates) [reply tests/test_rendering.py]: I didn't realize `st` was global, but looking at the fact that it's just a renamed import now that makes sense.

So sure in that case I think it makes sense to not do it that way.

File: example/schema.avsc (hunk 3 above)
Comment (jcoates) [reply example/schema.avsc]: Huh I didn't notice that there was a generated section checked in. If this stayed changed I could regenerate that.

I'm not sure I agree with the pitch that we're incompatible with the new version of avro, there's nothing special we're doing that is broken, it's the new version of avro that is broken for anyone who writes using the datum writer. We just happen to have a test that exposes that failure due to it's very generically named fields of `id` and `value`.

I don't think this is a bad solution any more than pinning the dependency into the past to avoid breaking a test is, but if you really do strongly prefer it I can do it. I checked and it seems like the bigger changes we're missing out on are:
 - support for UUIDs
 - allowing namespaces named `""`
 - validation is no longer implemented recursively and provides better error messages (this is the change that breaks this test)
 - fixing some imports that were causing conflicts with other dependencies

So that's the tradeoff, but nothing there sounds critical.

File: avro2py/codegen.py (hunk 2 above)
Comment (jcoates) [reply avro2py/codegen.py]: These are passed in and it's static again.

File: example/schema.avsc (hunk 3 above)
Comment (jcoates) [reply example/schema.avsc]: I pinned the version and reverted the change.

File: avro2py/codegen.py (hunk 4 above)
Comment (jcoates) [reply avro2py/codegen.py]: This logic got changed.

File: avro2py/codegen.py (hunk 5 above)
Comment (jcoates) [reply avro2py/codegen.py]: I used zip for this in the replacement version.

File: avro2py/codegen.py (hunk 4 above)
Comment (jcoates) [reply avro2py/codegen.py]: this is closer to what happens now (although not exactly).

File: avro2py/codegen.py (hunk 6 above)
Comment (jcoates) [reply avro2py/codegen.py]: no more += 1

File: avro2py/codegen.py (hunk 2 above)
Comment (jcoates) [reply avro2py/codegen.py]: removed

File: avro2py/codegen.py [hunk 7]
```diff
@@ -316,6 +316,21 @@ def resolve_field_type(field_type: Union[AvroPrimitives, DefinedType, Record, En
...
+    A parent namespace is a namespace which is a prefix.
+
+    For example, `a.b` is a parent of `a.b.c.d`.
+    """
+    for parent_part, child_part in zip(potential_parent, potential_child):
+        if parent_part != child_part:
+            return False
+    if len(potential_parent) >= len(potential_child):
+        return False
+    return True
```
Comment (hchasestevens) [inline avro2py/codegen.py]:
```suggestion
    return len(potential_parent) < len(potential_child)
```

File: avro2py/codegen.py [hunk 8]
```diff
@@ -316,6 +316,21 @@ def resolve_field_type(field_type: Union[AvroPrimitives, DefinedType, Record, En
...
+def namespace_is_parent_of(potential_parent: List[str], potential_child: List[str]) -> bool:
+    """Checks if one namespace is a parent of another.
+
+    A parent namespace is a namespace which is a prefix.
+
+    For example, `a.b` is a parent of `a.b.c.d`.
+    """
+    for parent_part, child_part in zip(potential_parent, potential_child):
+        if parent_part != child_part:
+            return False
```
Comment (hchasestevens) [inline avro2py/codegen.py]:
```suggestion
    if any(parent_part != child_part for parent_part, child_part in zip(potential_parent, potential_child)):
        return False
```

File: avro2py/codegen.py [hunk 9]
```diff
@@ -338,9 +353,24 @@ def rewriter(self, module_namespace: str, module_imports: List[Union[ast.Import,
...
+    def __init__(self, namespaces: Dict[str, Tuple[ast.AST, List[Union[ast.Import, ast.ImportFrom]]]]):
+        super(RewriteCrossReferenceStrings, self).__init__(namespaces)
+
+        namespace_parts = [namespace.split(".") for namespace in namespaces]
+        namespaces_with_children = [
+            ".".join(potential_parent)
+            for potential_parent, potential_child in itertools.permutations(namespace_parts, 2)
+            if namespace_is_parent_of(potential_parent, potential_child)
+        ]
+        self.namespaces_with_children = frozenset(namespaces_with_children)
```
Comment (hchasestevens) [inline avro2py/codegen.py]:
```suggestion
        self.namespaces_with_children = {
            ".".join(potential_parent)
            for potential_parent, potential_child in itertools.permutations(namespace_parts, 2)
            if namespace_is_parent_of(potential_parent, potential_child)
        }
```
so we just make the set directly from the get-go

File: avro2py/codegen.py [hunk 10]
```diff
@@ -338,9 +353,24 @@ def rewriter(self, module_namespace: str, module_imports: List[Union[ast.Import,
 class RewriteCrossReferenceStrings(ModuleAwareNodeTransformer):
     """Used to rewrite deferred type annotations within an AnnAssign node."""
 
+    def __init__(self, namespaces: Dict[str, Tuple[ast.AST, List[Union[ast.Import, ast.ImportFrom]]]]):
+        super(RewriteCrossReferenceStrings, self).__init__(namespaces)
+
+        namespace_parts = [namespace.split(".") for namespace in namespaces]
+        namespaces_with_children = [
+            ".".join(potential_parent)
+            for potential_parent, potential_child in itertools.permutations(namespace_parts, 2)
```
Comment (hchasestevens) [inline avro2py/codegen.py:359]: This being n^2 instead of doing something like creating a tree or sorting and then comparing elements pairwise feels inelegant, but I guess it doesn't really matter.

File: avro2py/codegen.py [hunk 11]
```diff
@@ -363,10 +393,15 @@ def bridge_namespaces(from_namespace: List[str], to_namespace: List[str], target
         remaining_to_components = [c for c in paired_remaining_to_components if c is not None]
 
         if remaining_from_components:
+            from_namespace_as_str = '.'.join(from_namespace)
+            if from_namespace_as_str in namespaces_with_children:
+                level = len(remaining_from_components) + 1
```
Comment (hchasestevens) [inline avro2py/codegen.py:396]: Can you add a comment here explaining why we do this?

File: avro2py/codegen.py [hunk 12]
```diff
@@ -363,10 +393,15 @@ def bridge_namespaces(from_namespace: List[str], to_namespace: List[str], target
         remaining_to_components = [c for c in paired_remaining_to_components if c is not None]
 
         if remaining_from_components:
+            from_namespace_as_str = '.'.join(from_namespace)
+            if from_namespace_as_str in namespaces_with_children:
```
Comment (hchasestevens) [inline avro2py/codegen.py]: Rather than having to do `'.'.join` on all of these, why don't we just _not_ do the `".".join` in https://github.com/hchasestevens/avro2py/pull/15/files#diff-dd17a87ea1ee28b0a8ef80f33739b424462dfd309d0a7eb4d58b619f0ec386f3R361 and have these be `tuple` instead of `list` (so they're hashable and will work as set members)? I think list->tuple conversion should be much faster than string joining.

**Review by hchasestevens — APPROVED** [review]
Left a few comments on potential improvements, but overall looks like it fixes the problem. Thanks for the contribution.

File: tests/test_rendering.py [hunk 13]
```diff
@@ -50,12 +50,20 @@ def test_example_message_round_trippable(data):
     tmp_path.unlink()
 
     # clean up PBT-generated example model
+    # we would like to focus on non-infinite, numeric decimals
+    # due to a bug in avro python implementation; exponent _must_ match scale
+    original_example_avro_model = original_example_avro_model._replace(
+        decimal=data.draw(st.decimals(allow_nan=False, allow_infinity=False, places=2))
+    )
+    if original_example_avro_model.maybeDecimal:
```
Comment (hchasestevens) [inline tests/test_rendering.py]:
```suggestion
    if original_example_avro_model.maybeDecimal is not None:
```
-- consider:
```python
In [1]: from decimal import Decimal

In [2]: if Decimal('0.00000000000'):
   ...:     print('foo')
   ...:
```

File: avro2py/codegen.py (hunk 11 above)
Comment (jcoates) [reply avro2py/codegen.py:396]: Yes I can, although I think the situation is pretty specific so I'm not sure how clear the comment will be.

File: avro2py/codegen.py (hunk 10 above)
Comment (jcoates) [reply avro2py/codegen.py:359]: I think it seems a little inefficient but relatively straightforward.

**Review by hchasestevens — APPROVED** [review]
